tests/src/Unit/PermissionEventTest.php

```diff
@@ -380,6 +547,73 @@ public function testIteratorAggregate($permissions, $entity_type_id, $bundle_id,
   }
 
   /**
+   * Tests creation of an invalid operation permission.
+   *
+   * @expectedException \InvalidArgumentException
+   */
+  public function testInvalidGroupContentOperationPermissionCreation() {
+    // An exception should be thrown when a group content operation permission
+    // is created with an invalid ownership type.
+    new GroupContentOperationPermission([
+      'name' => 'invalid permission',
+      'title' => $this->t('This is an invalid permission.'),
+      'entity type' => 'node',
+      'bundle' => 'article',
+      'operation' => 'create',
+      'ownership' => 'an invalid ownership',
+    ]);
+  }
+
+  /**
+   * Tests that it is possible to overwrite an existing operation permission.
+   *
+   * Operation permissions are not identified by their machine name, but by the
+   * unique combination of entity type, bundle, operation and ownership. It
+   * should be possible to change the name, title and other properties.
+   */
+  public function testOverwritingGroupContentOperationPermission() {
+    $event = new PermissionEvent($this->randomMachineName(), $this->randomMachineName(), []);
+
+    $entity_type_id = $this->randomMachineName();
+    $bundle_id = $this->randomMachineName();
+    $operation = $this->randomMachineName();
+    $ownership = 'own';
+
+    $original_permission = new GroupContentOperationPermission([
+      'name' => 'the original permission',
+      'title' => $this->t('This is the original permission.'),
+      'description' => $this->t('This is the original description.'),
+      'entity type' => $entity_type_id,
+      'bundle' => $bundle_id,
+      'operation' => $operation,
+      'ownership' => $ownership,
+      'default roles' => [OgRoleInterface::ADMINISTRATOR],
+      'restrict access' => TRUE,
+    ]);
+
+    $altered_permission = new GroupContentOperationPermission([
+      'name' => 'the altered permission',
+      'title' => $this->t('This is the altered permission.'),
+      'description' => $this->t('This is the altered description.'),
+      'entity type' => $entity_type_id,
+      'bundle' => $bundle_id,
+      'operation' => $operation,
+      'ownership' => $ownership,
+      'default roles' => [OgRoleInterface::AUTHENTICATED],
+      'restrict access' => FALSE,
+    ]);
+
+    // Check that the original permission can be set and retrieved.
+    $event->setPermission($original_permission);
+    $this->assertEquals($original_permission, $event->getGroupContentOperationPermission($entity_type_id, $bundle_id, $operation, $ownership));
+
+    // Check that the altered permission can be set and retrieved, even though
```

Comment:
The use case for this is to be able to support the arbitrary names that are used for entity operation permissions in core and contrib, for example `edit own article content` - this is actually a human readable name disguised as a machine name. If it was a real machine name it would be `update own article node`.

This makes the machine names unreliable to use as identifiers for a particular permission. Instead we're using the properties `entity type`, `bundle`, `operation` and `ownership` as identifiers, and it is possible to change the 'human readable machine name' at will. This is used for example to set the permissions for the node module in `OgEventSubscriber::provideDefaultNodePermissions()` - this changes the default permission machine name from `update own article node` to `edit own article content`.

It would work fine in the UI since these machine names are never shown to the end user. Instead we should the "real human readable name".

The whole thing is in fact arbitrary, the whole thing would work perfectly fine with a fixed machine name. But then we have to accept the fact that the machine names used in OG are not the same as the ones used in core + contrib. I added this mechanism so developers can use the permission names they are familiar with.